### PR TITLE
Add unit tests for API exceptions

### LINKED ISSUE #343: Add tests for API exceptions
Add tests for the exceptions added in PR #334

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes #343 

These tests mimic selection of appropriate exception for an API error response with a specific HTTP status code. All possible status codes from `System.Net.HttpStatusCode` are tested and some more integer values are tested as well.

Squash whenever.

### CODE REVIEW

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            [InlineData(550)]
+            [InlineData(599)]
+            public void ReturnsUnknownApiError(HttpStatusCode httpStatusCode)
+            {
+                var response = createErrorResponse(httpStatusCode);
+
+                var exception = ApiExceptions.ForResponse(response);
+
+                exception.Should().BeOfType<UnknownApiErrorException>();
+                ((UnknownApiErrorException) exception).HttpCode.Should().Equals(httpStatusCode);
```
Comment (amulware) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: Instead, you should be able to do `BeOfType<Blah>().Which.HttpCode.Should()..`

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            [InlineData(UnsupportedMediaType)]
+            [InlineData(RequestedRangeNotSatisfiable)]
+            [InlineData(ExpectationFailed)]
+            [InlineData(450)]
+            [InlineData(499)]
+            [InlineData(506)]
+            [InlineData(507)]
+            [InlineData(550)]
+            [InlineData(599)]
+            public void ReturnsUnknownApiError(HttpStatusCode httpStatusCode)
```
Comment (willsb) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: This is a strong candidate to be a property test :)

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            [InlineData(ServiceUnavailable, typeof(ServiceUnavailableException))]
+            [InlineData(GatewayTimeout, typeof(GatewayTimeoutException))]
+            [InlineData(HttpVersionNotSupported, typeof(HttpVersionNotSupportedException))]
+            public void ReturnsServerErrorException(HttpStatusCode httpStatusCode, Type expectedExceptionType)
+            {
+                var response = createErrorResponse(httpStatusCode);
+
+                var exception = ApiExceptions.ForResponse(response);
+
+                exception.Should().BeAssignableTo<ServerErrorException>();
```
Comment (willsb) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: You can `.And` here to chain the assertions

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            [InlineData(RequestEntityTooLarge, typeof(RequestEntityTooLargeException))]
+            [InlineData(418, typeof(ClientDeprecatedException))] // HTTP 418 - I Am a Teapot
+            [InlineData(429, typeof(TooManyRequestsException))] // HTTP 429 - Too Many Requests
+            public void ReturnsClientErrorException(HttpStatusCode httpStatusCode, Type expectedExceptionType)
+            {
+                var response = createErrorResponse(httpStatusCode);
+
+                var exception = ApiExceptions.ForResponse(response);
+
+                exception.Should().BeAssignableTo<ClientErrorException>();
```
Comment (willsb) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: You can `.And` here to chain the assertions

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs (hunk 2 above)
Comment (amulware) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: If changed to a property test, either use an int parameter and cast in the method or please debug and make sure the framework is able to generate random and non-existent enum values.

Comment (simonrozsival) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: I am not sure if this should be a property test - I still have to put the list of the known client and server errors somewhere and then go from 400 to 599 and exclude the ones we know - I could just use `UserData` instead for all of the three tests.

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs [hunk 5]
```diff
@@ -89,5 +56,52 @@ public void ReturnsUnknownApiError(HttpStatusCode httpStatusCode)
...
+                new object[] { NotImplemented, typeof(NotImplementedException) },
+                new object[] { BadGateway, typeof(BadGatewayException) },
+                new object[] { ServiceUnavailable, typeof(ServiceUnavailableException) },
+                new object[] { GatewayTimeout, typeof(GatewayTimeoutException) },
+                new object[] { HttpVersionNotSupported, typeof(HttpVersionNotSupportedException) }
+            };
+
+        public static bool IsKnownError(int code)
+        {
+            return ClientErrorsList.Concat(ServerErrorsList).FirstOrDefault(item => (int)item[0] == code) != null;
```
Comment (amulware) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: `Contains()` not working?

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs [hunk 6]
```diff
@@ -89,5 +56,52 @@ public void ReturnsUnknownApiError(HttpStatusCode httpStatusCode)
...
+        }
+
+        public static IEnumerable<object[]> UnknownErrorsList()
+        {
+            for (int i = 400; i < 600; ++i)
```
Comment (amulware) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: Replace with `Enumerable.Range()`?

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs (hunk 5 above)
Comment (simonrozsival) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: `Contains` does not take a predicate as parameter, but `Any` does the job.

File: Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs (hunk 6 above)
Comment (simonrozsival) [inline Toggl.Ultrawave.Tests/Exceptions/ApiErrorResponsesTests.cs]: Good idea.

**Pull request merged**